# Исключения

* ArithmeticException - потомок RuntimeException, возникает при вычислительных ошибках.

* OutOfMemoryError - возникает при переполнении памяти, потомок VirtualMachineError

* StackOverflowError - возникает при переполенни стека вызовов. Потомок VirtualMachineError.

* NullPointerException - возникает при работе с объектной переменной (указателем), который не указывает на реальный объект в памяти. Потомок RuntimeException.

* FileNotFoundException - возникает, когда система не может найти файл. Потомок IOException.

* IOException - все ошибки ввода-вывода. Потомок Exception.

* VirtualMachineError - ошибки виртуальной машины. Потомок Error.

* Error - ошибки, являющиеся критичиескими для программы, потомок Throwable

* RuntimeException - ?, потомок Exception

* Exception - ?, потомок Throwable

* Throwable - класс-предок всех исключений в Java.

```
private String detailMessage; // подробная информация об исключении

private static final StackTraceElement[]; // массив элементов стека вызовов

private Throwable cause; // причина текущего исключения
```

* StackTraceElement - элемент стека вызовов

```
	private String declaringClass;
    private String methodName;
    private String fileName;
    private int    lineNumber;
```

* Проверяемые исключения (checked exceptions) - Throwable, Exception и все его потомки, за исключением RuntimeException. Если программа "потенциально может выбросить" такое исключение, вы ПРЯМ ВОТ ВООБЩЕ ОБЯЗОН обработать его в коде.

* Обработка исключения - написание кода, который сработает, если исключение было проброшено.

```
try {
	// блок кода, в котором потенциально может возникнуть исключение
} catch (КлассИсключения объектИсключения) {
	// инструкции при возникновении ошибки
}
```

* throws после названия метода - означает, что ваш метод потенциально может выбросить какое-либо исключение, и возможно (если исключение проверяемое) вызывающий код должен будет его обработать, либо так же пробросить наверх.

* throw new - выброс исключение (можно выбрасывать все, что Throwable). Если исключение проверяемое, вы обязаны написать throws. (Вы можете и сами его обработать, но это дичь какая-то).

## Best Practictes
1. Обрабатывать проверяемое исключение "на месте", не пробрасывая его наверх.
2. НИКОГДА НЕ ИСПОЛЬЗОВАТЬ e.printStackTrace();
3. Пробрасывать вместо проверяемого непроверяемое.

## ЗАЧЕМ?

1. Может дойти до того, что клиенту в браузере будет выведена структура вашей базы данных или файлового хранилища, а это плохо.
2. Программа продолжит выполнятся, несмотря на возникшую критическую ситуацию.
3. Чтобы поток перестал выполняться, при этом клиентскому коду теперь нет необходимости обрабатывать ваши исключения.